File: frontend/src/app/core/services/dokumen-karyawan.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import {
  DokumenKaryawan, DokumenKaryawanRequest, KuotaDokumen
} from '../models/dokumen-karyawan.model';

@Injectable({ providedIn: 'root' })
export class DokumenKaryawanService {
  private http = inject(HttpClient);
  private url  = (karyawanId: number) =>
    `${environment.apiUrl}/karyawan/${karyawanId}/dokumen`;

  getAll(karyawanId: number, search?: string): Observable<DokumenKaryawan[]> {
    let params = new HttpParams();
    if (search) params = params.set('search', search);
    return this.http.get<any>(this.url(karyawanId), { params })
      .pipe(map(d => d?.dokumen ?? []));
  }

  getKuota(karyawanId: number): Observable<KuotaDokumen> {
    return this.http.get<any>(`${this.url(karyawanId)}/kuota`)
      .pipe(map(d => d.kuota));
  }

  create(karyawanId: number, req: DokumenKaryawanRequest, berkas: File): Observable<DokumenKaryawan> {
    const fd = new FormData();
    fd.append('data', new Blob([JSON.stringify(req)], { type: 'application/json' }));
    fd.append('berkas', berkas);
    return this.http.post<any>(this.url(karyawanId), fd).pipe(map(d => d.dokumen));
  }

  update(karyawanId: number, id: number, req: DokumenKaryawanRequest, berkas?: File): Observable<DokumenKaryawan> {
    const fd = new FormData();
    fd.append('data', new Blob([JSON.stringify(req)], { type: 'application/json' }));
    if (berkas) fd.append('berkas', berkas);
    return this.http.put<any>(`${this.url(karyawanId)}/${id}`, fd).pipe(map(d => d.dokumen));
  }

  delete(karyawanId: number, id: number): Observable<void> {
    return this.http.delete<void>(`${this.url(karyawanId)}/${id}`);
  }

  tambahKuotaBerbayar(karyawanId: number, jumlah: number): Observable<any> {
    return this.http.post<any>(`${this.url(karyawanId)}/kuota-berbayar`, { jumlah });
  }
}

File: backend/src/main/java/id/lua/absensikerja/service/DokumenKaryawanService.java
package id.lua.absensikerja.service;

import id.lua.absensikerja.model.dto.DokumenKaryawanDTO;
import id.lua.absensikerja.model.dto.DokumenKaryawanRequest;
import id.lua.absensikerja.model.dto.KuotaDokumenDTO;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

public interface DokumenKaryawanService {

    List<DokumenKaryawanDTO> findByKaryawan(Long karyawanId, String search);

    DokumenKaryawanDTO findById(Long id);

    DokumenKaryawanDTO create(Long karyawanId, DokumenKaryawanRequest req, MultipartFile berkas) throws IOException;

    DokumenKaryawanDTO update(Long id, DokumenKaryawanRequest req, MultipartFile berkas) throws IOException;

    void delete(Long id);

    KuotaDokumenDTO getKuota(Long karyawanId);

    KuotaDokumenDTO tambahKuotaBerbayar(Long karyawanId, int jumlah);
}
